File: client/models/Equip.ts
import Member from "./Member";

export interface EquipDraft {
  name: string;
  for_practice: boolean;
  for_game: boolean;
  description: string;
  storage_type: string; // "warehouse" | "takehome" | ""
}

export class Custody {
  constructor(
    public member_id: string,
    public timestamp: number,// t.Unix() * 1000
    public comment: string = "",
    public member: Member = null,
  ) { }
}

export default class Equip {
  constructor(
    public id: number,
    public name: string,
    public forPractice: boolean,
    public forGame: boolean,
    public description: string = "",
    public history: Custody[] = [],
    public storageType: string = "",
  ) { }

  static fromAPIResponse({ id, key, name, for_practice, for_game, description, history, storage_type }): Equip {
    return new Equip(id, name, for_practice, for_game, description, history ?? [], storage_type ?? "");
  }
  static listFromAPIResponse(res: { id, key, name, for_practice, for_game, description, history, storage_type }[]): Equip[] {
    return res.map(Equip.fromAPIResponse);
  }

  static draft(equip?: Equip): EquipDraft  {
    if (equip) {
      return { name: equip.name, for_practice: equip.forPractice, for_game: equip.forGame, description: equip.description, storage_type: equip.storageType ?? "" };
    }
    return { name: "", for_practice: false, for_game: false, description: "", storage_type: "" };
  }

  static sort(p: Equip, n: Equip): 1|-1 {
    return p.name < n.name ? -1 : 1;
  }
}

File: server/models/equip.go
package models

import (
	"regexp"
	"strings"
	"time"

	"cloud.google.com/go/datastore"
)

type (
	Equip struct {
		ID          int64          `json:"id" datastore:"-"`
		Key         *datastore.Key `datastore:"__key__"`
		Name        string         `json:"name"`
		ForPractice bool           `json:"for_practice"`
		ForGame     bool           `json:"for_game"`
		Description string         `json:"description"`
		History     []Custody      `json:"history" datastore:"-"`
	}

	Custody struct {
		Key       *datastore.Key `datastore:"__key__"`
		MemberID  string         `json:"member_id"`
		Timestamp int64          `json:"ts"`
		Comment   string         `json:"comment"`
	}
)

var (
	EquipmentExceptionExpression = regexp.MustCompile("!\\(([^)]+)\\)")
)

func (equip Equip) NeedsCharge() bool {
	return strings.HasPrefix(equip.Name, "ビデオ")
}

func (equip Equip) HasBeenUpdatedSince(t time.Time) bool {
	if len(equip.History) == 0 {
		return false
	}
	lastUpdated := time.Unix(equip.History[0].Timestamp/1000, 0)
	return lastUpdated.After(t)
}

func (equip Equip) ShouldBringFor(event Event) bool {
	if !equip.ForPractice && !equip.ForGame {
		return false
	}
	if event.IsGame() && equip.ForGame {
		return true
	}
	if !event.IsPractice() || !equip.ForPractice {
		return false
	}
	// 以下、練習用の特殊ケース
	match := EquipmentExceptionExpression.FindStringSubmatch(equip.Description)
	if len(match) < 2 {
		return equip.ForPractice
	}
	for _, exception := range strings.Split(match[1], ",") {
		if strings.Contains(event.Google.Title, exception) {
			return false
		}
	}
	return equip.ForPractice
}
